src/wallet/mod.rs

```diff
@@ -1339,23 +1359,42 @@ where
             return Ok((must_spend, vec![]));
         }
 
-        let satisfies_confirmed = match must_only_use_confirmed_tx {
-            true => {
-                let database = self.database.borrow();
-                may_spend
-                    .iter()
-                    .map(|u| {
-                        database
-                            .get_tx(&u.0.outpoint.txid, true)
-                            .map(|tx| match tx {
-                                None => false,
-                                Some(tx) => tx.confirmation_time.is_some(),
-                            })
+        let database = self.database.borrow();
+        let satisfies_confirmed = may_spend
+            .iter()
+            .map(|u| {
+                database
+                    .get_tx(&u.0.outpoint.txid, true)
+                    .map(|tx| match tx {
+                        // We don't have the tx in the db for some reason,
+                        // so we can't know for sure if it's mature or not.
+                        // We prefer not to spend it.
+                        None => false,
+                        Some(tx) => {
+                            let mut confirmed = true;
+                            if tx
+                                .transaction
+                                .expect("We specifically ask for the transaction above")
+                                .is_coin_base()
```

Comment:
I feel like this part of the code can be simplified a little?? 

Instead of storing `satisfied_confirmed` in a `Vec<bool>` then removing the `may_spend` utxos as per the index of those satisfactions, why not just remove all the immature coinbase utxos from the `may_spend`  in just one single list operation?? Could make the intention of the logic much clearer..

Also what happens when the user manually provides and immature coinbase??

I think the nicest way would be to 
 - have lambda that checks for maturity given an utxo..
 - List out `may_spend` and `must_spend` where that maturity satisfaction fails.
 -  Error early if such failure occures in user given `must_spend`.

Reply:
> Instead of storing satisfied_confirmed in a Vec<bool> then removing the may_spend utxos as per the index of those satisfactions, why not just remove all the immature coinbase utxos from the may_spend in just one single list operation?? Could make the intention of the logic much clearer..

It was mostly because I didn't want to modify the old `satisfies_confirmed` logic but sure, I can change it 

> Also what happens when the user manually provides and immature coinbase??

I strongly disagree with removing the immature coinbase inputs from the `must_spend`, as the list is intended for the utxos that we **must** spend. The API is intended for advanced users that know what they need - and it's up to them to check if that certain utxo can even be spent (we don't even check if it's spent already or not, for example).

Reply:
> as the list is intended for the utxos that we must spend. 

I agree with that.. Just a flow thought that came while figuring the list operations.. Yes we should not remove it from there if that's the API intention..

Reply:
> But then again advanced users "not knowing" what they are doing, and spending an immature coinbase would never get them into blockchain as it is consensus enforced.. Will not cause fund loss though.. But no hard opinions one way or other.. If someone wanna do that, might have some good reason for it..

I think it's reasonable to assume that whoever's using an advance API knows what they're doing. Also, they might want to presign a transaction and then hold it for a while, and so they wouldn't care about coinbase maturity.

> Another way would be to modify the get_available_utxos() function, to not return immature coinbase.. This function is only used in may_spend extraction, and nowhere else.. So might make sense to enforce the behavior at that level itself..

That would probably be slightly clearer, but it would require an additional db call (now we retrieve the tx once, and we use it to check if the tx is confirmed and if the inputs are mature)

What do you think of this? Is it slightly more readable? I'm sadly discarding every error `get_tx` throws at me, as I can't have `?` inside retain
```diff
         let database = self.database.borrow();
-        let satisfies_confirmed = may_spend
-            .iter()
-            .map(|u| {
-                database
-                    .get_tx(&u.0.outpoint.txid, true)
-                    .map(|tx| match tx {
-                        // We don't have the tx in the db for some reason,
-                        // so we can't know for sure if it's mature or not.
-                        // We prefer not to spend it.
-                        None => false,
-                        Some(tx) => {
-                            let mut confirmed = true;
-                            if tx
-                                .transaction
-                                .expect("We specifically ask for the transaction above")
-                                .is_coin_base()
-                            {
-                                if let Some(current_height) = current_height {
-                                    if let Some(tx_conf_time) = &tx.confirmation_time {
-                                        // https://github.com/bitcoin/bitcoin/blob/c5e67be03bb06a5d7885c55db1f016fbf2333fe3/src/validation.cpp#L373-L375
-                                        confirmed &= (current_height - tx_conf_time.height)
-                                            >= COINBASE_MATURITY;
-                                    } else {
-                                        confirmed = false;
-                                    }
-                                }
-                            }
-                            if must_only_use_confirmed_tx {
-                                confirmed &= tx.confirmation_time.is_some();
+
+        let satisfies_confirmed = |u: &LocalUtxo| {
+            database
+                .get_tx(&u.outpoint.txid, true)
+                .ok()
+                .flatten()
+                .map(|tx| {
+                    let mut confirmed = true;
+                    if tx
+                        .transaction
+                        .expect("We specifically ask for the transaction above")
+                        .is_coin_base()
+                    {
+                        if let Some(current_height) = current_height {
+                            if let Some(tx_conf_time) = &tx.confirmation_time {
+                                // https://github.com/bitcoin/bitcoin/blob/c5e67be03bb06a5d7885c55db1f016fbf2333fe3/src/validation.cpp#L373-L375
+                                confirmed &=
+                                    (current_height - tx_conf_time.height) >= COINBASE_MATURITY;
+                            } else {
+                                confirmed = false;
                             }
-                            confirmed
                         }
-                    })
-            })
-            .collect::<Result<Vec<_>, _>>()?;
+                    }
+                    if must_only_use_confirmed_tx {
+                        confirmed &= tx.confirmation_time.is_some();
+                    }
+                    confirmed
+                })
+                .unwrap_or(false)
+        };
 
         let mut i = 0;
         may_spend.retain(|u| {
             let retain = change_policy.is_satisfied_by(&u.0)
                 && !unspendable.contains(&u.0.outpoint)
-                && satisfies_confirmed[i];
+                && satisfies_confirmed(&u.0);
             i += 1;
             retain
         });

Reply:
I feel for now its ok to leave it as it.. Maybe in a later PR we can move this all together in `get_available_utxos()` and simplify this logic a bit..